### ISSUE TITLE
mmap error with maven

### ISSUE THREAD

**Issue by IordanisSap** [eellak/build-recorder#167]
While building a java project with maven i encountered the following issue
```
./build_recorder: mmaping `/sys/devices/system/cpu/online': No such device
Segmentation fault (core dumped)
```
It probably has to do with files that do not support memory mapping.

**Comment by zvr** [comment]
OK, that's...  interesting.

According to https://github.com/eellak/build-recorder/blob/22976965759cb293b96bb4a868e24ee706ac58c5/src/hash.c#L65 the case where `mmap(2)` fails is already being handled.

**Comment by fvalasiad** [comment]
@zvr The issue is with `mmap(2)`. There are system files that can be `open(2)`ed and `read(2)` as expected but not `mmap(2)`ed and this is why this fails.

In the future we need to remove mmap anyways to perform an important optimization. That is the underlying tracee only needs to be stopped for the duration which we read the file, they don't have to wait for the hashing too. So perhaps we should get rid of it here as well.

Instead we will just read the file into a buffer, this might result in some additional space complexity but it shouldn't be an issue.

This bug will also disappear that way.

**Comment by zvr** [comment]
@fvalasiad I'm not sure I understand. Does `mmap(2)` fail or not? And if it does, doesn't the abovementioned line handle this?

**Comment by fvalasiad** [comment]
@zvr Yes, `mmap(2)` fails and the above line goes into the if statement producing the error code @IordanisSap mentions in that issue.

Are you asking why do we get the segfault since inside that if statement we return a NULL?

**Comment by zvr** [comment]
Yes, the `mmap` failure is correctly handled: fie is not mapped, so NULL is returned. No problem there.

Why crash, then?

**Comment by fvalasiad** [comment]
@zvr `hash_file_contents(3)` returns a hash which is fed to `hash_to_str(3)` which expects a not-null buffer. Thus the segfault.

I am arguing though that there are files which their contents are something we care for but we can't `mmap(2)` them. So just returning a null is incorrect.

**Comment by fvalasiad** [comment]
This should do it for now. I tested with `./build-recorder lsblk` which also crashed for the same reason and it now works as expected.
